components/brave_wallet/browser/brave_wallet_pin_service.cc

```diff
@@ -0,0 +1,778 @@
+}
+
+// static
+absl::optional<std::string> BraveWalletPinService::GetPath(
+    const absl::optional<std::string>& service,
+    const BlockchainTokenPtr& token) {
+  if (service && base::ContainsOnlyChars(service.value(), ".")) {
+    return absl::nullopt;
+  }
+  if (base::ContainsOnlyChars(token->contract_address, ".")) {
+    return absl::nullopt;
+  }
+  if (base::ContainsOnlyChars(token->token_id, ".")) {
+    return absl::nullopt;
+  }
+  DCHECK(!base::ContainsOnlyChars(token->chain_id, "."));
+  return base::StrCat({kNftPart, ".", service.value_or(kLocalService), ".",
+                       base::NumberToString(static_cast<int>(token->coin)), ".",
+                       token->chain_id, ".", token->contract_address, ".",
+                       token->token_id});
+}
+
+// static
+BlockchainTokenPtr BraveWalletPinService::TokenFromPath(
+    const std::string& path) {
+  std::vector<std::string> parts =
+      base::SplitString(path, ".", base::TRIM_WHITESPACE, base::SPLIT_WANT_ALL);
+  if (parts.size() != 6) {
+    return nullptr;
+  }
+  mojom::BlockchainTokenPtr token = mojom::BlockchainToken::New();
+  int32_t coin;
+  if (!base::StringToInt(parts.at(2), &coin)) {
+    return nullptr;
+  }
+  token->coin = static_cast<mojom::CoinType>(coin);
+  token->chain_id = parts.at(3);
+  token->contract_address = parts.at(4);
+  token->token_id = parts.at(5);
+  token->is_erc721 = true;
+  token->is_nft = true;
+  return token;
+}
+
+// static
+absl::optional<std::string> BraveWalletPinService::ServiceFromPath(
+    const std::string& path) {
+  std::vector<std::string> parts =
+      base::SplitString(path, ".", base::TRIM_WHITESPACE, base::SPLIT_WANT_ALL);
+  if (parts.size() != 6) {
+    return nullptr;
+  }
+  if (parts.at(1) == kLocalService) {
+    return absl::nullopt;
+  } else {
+    return parts.at(1);
+  }
+}
+
+void BraveWalletPinService::Validate(BlockchainTokenPtr token,
+                                     const absl::optional<std::string>& service,
+                                     ValidateCallback callback) {
+  mojom::TokenPinStatusPtr status = GetTokenStatus(service, token);
+  if (!status) {
+    std::move(callback).Run(false, nullptr);
+    return;
+  }
+  if (status->code != mojom::TokenPinStatusCode::STATUS_PINNED) {
+    std::move(callback).Run(false, nullptr);
+    return;
+  }
+
+  absl::optional<std::vector<std::string>> cids =
+      ResolvePinItems(service, token);
+
+  if (!cids) {
+    SetTokenStatus(service, std::move(token),
+                   mojom::TokenPinStatusCode::STATUS_PINNING_IN_PROGRESS,
+                   nullptr);
+    std::move(callback).Run(true, nullptr);
+    return;
+  }
+
+  auto path = GetPath(absl::nullopt, token);
+  if (!path) {
+    std::move(callback).Run(
+        false,
+        mojom::PinError::New(mojom::WalletPinServiceErrorCode::ERR_WRONG_TOKEN,
+                             "Wrong token data"));
+    return;
+  }
+
+  if (!service) {
```

Comment:
Again I'm a bit confused about why passing absl::nullopt as service means local.

Reply:
The initial interface was designed to support remote and local pinning services. So if service is remote we use non-nullopt service name, otherwise it should be null